src/storage/v2/durability/durability.cpp

```diff
@@ -434,13 +432,28 @@ std::optional<RecoveryInfo> Recovery::RecoverData(std::string *uuid, Replication
         recovery_info.next_timestamp = std::max(recovery_info.next_timestamp, info.next_timestamp);
 
         recovery_info.last_commit_timestamp = info.last_commit_timestamp;
+
+        if (recovery_info.next_timestamp != 0) {
+          last_loaded_timestamp.emplace(recovery_info.next_timestamp - 1);
+        }
+
+        bool epoch_history_empty = epoch_history->empty();
+        bool epoch_not_recorded = !epoch_history_empty && epoch_history->back().first != wal_file.epoch_id;
+        auto last_loaded_timestamp_value = last_loaded_timestamp.value_or(0);
+
+        if (epoch_history_empty || epoch_not_recorded) {
+          epoch_history->emplace_back(std::string(wal_file.epoch_id), last_loaded_timestamp_value);
+        }
+
+        auto last_epoch_updated = !epoch_history_empty && epoch_history->back().first == wal_file.epoch_id &&
+                                  epoch_history->back().second < last_loaded_timestamp_value;
+        if (last_epoch_updated) {
+          epoch_history->back().second = last_loaded_timestamp_value;
+        }
```

Comment:
```suggestion
        auto last_loaded_timestamp_value = last_loaded_timestamp.value_or(0);
        if (epoch_history->empty()) {
          // no history, add epoch
          epoch_history->emplace_back(wal_file.epoch_id, last_loaded_timestamp_value);
        } else if (epoch_history->back().first != wal_file.epoch_id) {
          // new epoch, add it
          epoch_history->emplace_back(wal_file.epoch_id, last_loaded_timestamp_value);
        } else if (epoch_history->back().second < last_loaded_timestamp_value) {
          // existing epoch, update with newer timestamp
          epoch_history->back().second = last_loaded_timestamp_value;
        }
```

Reply:
As discussed on call I will do it on next PR

Reply:
done